Fixes namespaced UUID generation for namespace IDs represented as strings

erichmachado opened the pull request:
### Summary

Updates the `Digest::UUID.uuid_from_hash` implementation to return the correct UUID value if the `uuid_namespace` parameter is provided as a `String` instead of using one of the pre-defined constants.

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->

### Other Information

The patch is backwards-compatible with the previous behaviour of using the constants defined in `Digest::UUID` as the namespace ID.

With the new behaviour, all the calls below should return the same value:

```ruby
require "digest"
require "active_support/core_ext/digest"

Digest::UUID.uuid_v5(Digest::UUID::DNS_NAMESPACE, "www.widgets.com")
=> "21f7f8de-8051-5b89-8680-0195ef798b6a"
Digest::UUID.uuid_v5("6BA7B810-9DAD-11D1-80B4-00C04FD430C8", "www.widgets.com")
=> "21f7f8de-8051-5b89-8680-0195ef798b6a"
Digest::UUID.uuid_v5("6ba7b810-9dad-11d1-80b4-00c04fd430c8", "www.widgets.com")
=> "21f7f8de-8051-5b89-8680-0195ef798b6a"
```

### Benchmarks
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "activesupport", "6.0.0"
  gem "benchmark-ips"
end

require "digest"
require "active_support/core_ext/digest"

# Your patch goes here.
module Digest
  module UUID
    def self.new_uuid_from_hash(hash_class, uuid_namespace, name)
      if hash_class == Digest::MD5
        version = 3
      elsif hash_class == Digest::SHA1
        version = 5
      else
        raise ArgumentError, "Expected Digest::SHA1 or Digest::MD5, got #{hash_class.name}."
      end

      packed_namespace_uuid = uuid_namespace.length == 36 ? pack_uuid(uuid_namespace) : uuid_namespace

      hash = hash_class.new
      hash.update(packed_namespace_uuid)
      hash.update(name)

      ary = hash.digest.unpack("NnnnnN")
      ary[2] = (ary[2] & 0x0FFF) | (version << 12)
      ary[3] = (ary[3] & 0x3FFF) | 0x8000

      "%08x-%04x-%04x-%04x-%04x%08x" % ary
    end

    def self.pack_uuid(uuid)
      uuid.scan(/(\h{8})-(\h{4})-(\h{4})-(\h{4})-(\h{4})(\h{8})/).flatten.map { |s| s.to_i(16) }.pack("NnnnnN")
    end
  end
end

# Enumerate some representative scenarios here.
#
# It is very easy to make an optimization that improves performance for a
# specific scenario you care about but regresses on other common cases.
# Therefore, you should test your change against a list of representative
# scenarios. Ideally, they should be based on real-world scenarios extracted
# from production applications.
SCENARIOS = {
  "Binary namespace ID"      => Digest::UUID::DNS_NAMESPACE,
  "Lower-cased namespace ID" => "6ba7b810-9dad-11d1-80b4-00c04fd430c8",
  "Upper-cased namespace ID" => "6BA7B810-9DAD-11D1-80B4-00C04FD430C8"
}

VALUE = "www.widgets.com"

SCENARIOS.each_pair do |name, uuid_namespace|
  puts
  puts " #{name} ".center(80, "=")
  puts

  Benchmark.ips do |x|
    x.report("old_uuid_from_hash_md5") { Digest::UUID.uuid_from_hash(Digest::MD5, uuid_namespace, VALUE) }
    x.report("new_uuid_from_hash_md5") { Digest::UUID.new_uuid_from_hash(Digest::MD5, uuid_namespace, VALUE) }
    x.compare!
  end

  Benchmark.ips do |x|
    x.report("old_uuid_from_hash_sha1") { Digest::UUID.uuid_from_hash(Digest::SHA1, uuid_namespace, VALUE) }
    x.report("new_uuid_from_hash_sha1") { Digest::UUID.new_uuid_from_hash(Digest::SHA1, uuid_namespace, VALUE) }
    x.compare!
  end
end
```

#### Results
```
ruby benchmark.rb
Fetching gem metadata from https://rubygems.org/..............
Resolving dependencies...
Using concurrent-ruby 1.1.5
Using i18n 1.7.0
Using minitest 5.13.0
Using thread_safe 0.3.6
Using tzinfo 1.2.5
Using zeitwerk 2.2.1
Using activesupport 6.0.0
Using benchmark-ips 2.7.2
Using bundler 2.0.2

============================= Binary namespace ID ==============================

Warming up --------------------------------------
old_uuid_from_hash_md5
                        14.605k i/100ms
new_uuid_from_hash_md5
                        14.328k i/100ms
Calculating -------------------------------------
old_uuid_from_hash_md5
                        166.129k (±10.3%) i/s -    832.485k in   5.074024s
new_uuid_from_hash_md5
                        163.218k (±10.5%) i/s -    816.696k in   5.080808s

Comparison:
old_uuid_from_hash_md5:   166128.7 i/s
new_uuid_from_hash_md5:   163218.5 i/s - same-ish: difference falls within error

Warming up --------------------------------------
old_uuid_from_hash_sha1
                        14.660k i/100ms
new_uuid_from_hash_sha1
                        14.050k i/100ms
Calculating -------------------------------------
old_uuid_from_hash_sha1
                        163.423k (±11.6%) i/s -    806.300k in   5.021001s
new_uuid_from_hash_sha1
                        159.354k (± 9.9%) i/s -    800.850k in   5.090466s

Comparison:
old_uuid_from_hash_sha1:   163422.6 i/s
new_uuid_from_hash_sha1:   159353.9 i/s - same-ish: difference falls within error


=========================== Lower-cased namespace ID ===========================

Warming up --------------------------------------
old_uuid_from_hash_md5
                        14.681k i/100ms
new_uuid_from_hash_md5
                         6.007k i/100ms
Calculating -------------------------------------
old_uuid_from_hash_md5
                        168.224k (± 8.2%) i/s -    836.817k in   5.012074s
new_uuid_from_hash_md5
                         68.417k (±19.1%) i/s -    324.378k in   5.033972s

Comparison:
old_uuid_from_hash_md5:   168224.3 i/s
new_uuid_from_hash_md5:    68417.4 i/s - 2.46x  slower

Warming up --------------------------------------
old_uuid_from_hash_sha1
                        14.321k i/100ms
new_uuid_from_hash_sha1
                         5.992k i/100ms
Calculating -------------------------------------
old_uuid_from_hash_sha1
                        165.424k (± 7.6%) i/s -    830.618k in   5.055094s
new_uuid_from_hash_sha1
                         67.677k (±19.0%) i/s -    323.568k in   5.070097s

Comparison:
old_uuid_from_hash_sha1:   165423.8 i/s
new_uuid_from_hash_sha1:    67676.6 i/s - 2.44x  slower


=========================== Upper-cased namespace ID ===========================

Warming up --------------------------------------
old_uuid_from_hash_md5
                        14.970k i/100ms
new_uuid_from_hash_md5
                         6.060k i/100ms
Calculating -------------------------------------
old_uuid_from_hash_md5
                        168.793k (± 8.5%) i/s -    838.320k in   5.012564s
new_uuid_from_hash_md5
                         68.246k (±19.0%) i/s -    327.240k in   5.082545s

Comparison:
old_uuid_from_hash_md5:   168793.3 i/s
new_uuid_from_hash_md5:    68245.5 i/s - 2.47x  slower

Warming up --------------------------------------
old_uuid_from_hash_sha1
                        14.683k i/100ms
new_uuid_from_hash_sha1
                         5.972k i/100ms
Calculating -------------------------------------
old_uuid_from_hash_sha1
                        164.037k (± 9.5%) i/s -    822.248k in   5.063807s
new_uuid_from_hash_sha1
                         67.914k (±18.9%) i/s -    322.488k in   5.033210s

Comparison:
old_uuid_from_hash_sha1:   164037.2 i/s
new_uuid_from_hash_sha1:    67913.9 i/s - 2.42x  slower
```

As can be seen from the benchmarks above, performance remains roughly the same for binary representations of namespace IDs, so no impact should be noted in this case.

For the `String` version of namespace IDs, there's a performance impact due to the extra work required to process them, but given that the current implementation returns the wrong value back in those cases, it seems like it's a fair trade-off.

Closes #37681

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

erichmachado committed:
Addresses `Digest::UUID` RFC 4122 compatibility.

Updates the `Digest::UUID.uuid_from_hash` to return the correct UUID values
for namespace IDs that are different from the ones provided on `Digest::UUID`.

- The new behavior will be enabled by setting the
  `config.active_support.use_rfc4122_namespaced_uuids` option to `true`
  and is the default for new apps.

- The old behavior is the default for upgraded apps and will output a
  deprecation warning every time a value that is different than one of
  the constants defined on the `Digest::UUID` extension is used as the
  namespace ID.

Fixes #37681.

erichmachado commented on activesupport/test/core_ext/digest/uuid_test.rb:
```diff
@@ -4,17 +4,46 @@
+
   def test_v3_uuids
+    assert_equal "3d813cbb-47fb-32ba-91df-831e1593ac29", Digest::UUID.uuid_v3("6BA7B810-9DAD-11D1-80B4-00C04FD430C8", "www.widgets.com")
+    assert_equal "3d813cbb-47fb-32ba-91df-831e1593ac29", Digest::UUID.uuid_v3("6ba7b810-9dad-11d1-80b4-00c04fd430c8", "www.widgets.com")
```
This is just testing that the values returned are the same for different representations of the same UUID. The same goes for the test cases below.

erichmachado commented on activesupport/test/core_ext/digest/uuid_test.rb:
```diff
@@ -4,17 +4,46 @@
+    assert_equal "6ba7b810-9dad-11d1-80b4-00c04fd430c8", "%08x-%04x-%04x-%04x-%04x%08x" % Digest::UUID::DNS_NAMESPACE.unpack("NnnnnN")
+    assert_equal "6ba7b811-9dad-11d1-80b4-00c04fd430c8", "%08x-%04x-%04x-%04x-%04x%08x" % Digest::UUID::URL_NAMESPACE.unpack("NnnnnN")
+    assert_equal "6ba7b812-9dad-11d1-80b4-00c04fd430c8", "%08x-%04x-%04x-%04x-%04x%08x" % Digest::UUID::OID_NAMESPACE.unpack("NnnnnN")
+    assert_equal "6ba7b814-9dad-11d1-80b4-00c04fd430c8", "%08x-%04x-%04x-%04x-%04x%08x" % Digest::UUID::X500_NAMESPACE.unpack("NnnnnN")
```
This ensures that the constants contain the values defined on the [Appendix C of RFC 4122](https://tools.ietf.org/html/rfc4122#appendix-C).

eugeneius commented on activesupport/lib/active_support/core_ext/digest/uuid.rb:
```diff
@@ -49,5 +51,11 @@ def self.uuid_v5(uuid_namespace, name)
     end
+
+    def self.pack_uuid(uuid)
+      uuid.scan(/(\h{8})-(\h{4})-(\h{4})-(\h{4})-(\h{4})(\h{8})/).flatten.map { |s| s.to_i(16) }.pack("NnnnnN")
```
This regex should be anchored so that it only accepts a UUID, rather than any string containing a UUID:

```suggestion
      uuid.match(/\A(\h{8})-(\h{4})-(\h{4})-(\h{4})-(\h{4})(\h{8})\z/).captures.map { |s| s.to_i(16) }.pack("NnnnnN")
```

erichmachado replied:
Hi @eugeneius. Thank you for your review. I've accepted the suggested change.

eugeneius commented:
Could you rebase and squash your commits? That should fix the test failures.

This now looks good to me, although it still needs a review from a committer.

agrobbin commented:
I ran into this issue a little while back, and *love* that it's being fixed @erichmachado @eugeneius, but I'm wondering if it should be done in multiple stages to avoid breaking existing uses of the utility. If someone is using this utility with a string representation of a UUID, and then upgrades their app to a version of Rails that includes this change, the generated UUIDs will not be consistent, which could have some really nasty unintended consequences for them.

```ruby
uuid_namespace = Digest::UUID.uuid_v5(Digest::UUID::DNS_NAMESPACE, "www.example.com")
# => "2ed6657d-e927-568b-95e1-2665a8aea6a2"

# before this change
Digest::UUID.uuid_v5("2ed6657d-e927-568b-95e1-2665a8aea6a2", "foo")
# => "c96ed0bd-aac0-5b5d-95b0-e21bf62334be"

# after this change
Digest::UUID.uuid_v5("2ed6657d-e927-568b-95e1-2665a8aea6a2", "foo")
# => "a50fdd48-b8e7-5669-b8c4-2b6f7218a728"
```

Would it make sense to make this change opt-in for 1 major release cycle, with a warning any time someone passes a string representation of a UUID? Then, when people are prepared to make the change, they can set a configuration to do so (with that configuration becoming the default in a future major release).

erichmachado commented:
I like the idea @agrobbin. I'm willing to implement the configuration option to support that, but I'm not sure how to do it. If anyone knows the right direction it would be of great help. I'll check the codebase as well to see if I can find any examples.

Any suggestions regarding the opt-in setting name are also welcome.

erichmachado commented:
Hi @agrobbin, I followed your suggestion and added a configuration option to enable the new behavior. The setting is `Rails.application.config.active_support.use_rfc4122_namespaced_uuids` and it defaults to `false` (for backwards compatibility).

If the new behavior is disabled, every time a string representation of a UUID is used as a namespace, a deprecation warning will output. Tests were updated to check the configuration option added and the `Digest::UUID` output for each setting.

I've also updated the `CHANGELOG.md` and the documentation. I've included your name on the contributors list (I hope you're OK with that).

agrobbin commented:
Nice @erichmachado! Can't wait to hear what Rails Core thinks!

erichmachado commented:
Hi @eugeneius, sorry to bother you again. I've recently rebased this branch and fixed the conflicts. I've also squashed the commits as requested and also added a commit message formatted according to the contribution guideline. Is there anything else needed on my end?

Quintasan commented:
@erichmachado @eugeneius Hi, I've been hit by this recently and I was wondering if we could get the ball rolling on this

erichmachado commented:
> @erichmachado @eugeneius Hi, I've been hit by this recently and I was wondering if we could get the ball rolling on this

Sure, I can take care of the conflicts if you're interested in merging this.

Quintasan commented:
I'm interested in getting this merged but I have no commit rights. I can pitch in as long as someone from Rails team expresses interest in getting this merged

erichmachado commented:
Hi @Quintasan,

I've updated the branch to fix the conflict and also added one more change to ensure only UUIDs are accepted as namespace identifiers if RFC 4122 compatibility is enabled (an `ArgumentError` should be raised otherwise). Previous behaviour is not changed and I've added extra specs to ensure that.

The build has failed but it seems unrelated.

/cc @eugeneius

Quintasan commented:
@erichmachado thanks for your efforts!

Sorry to select you out @eileencodes but I have seen you being active in the PRs so I thought maybe you could take a look at this and point out any things we should change to get this merged.

eileencodes commented:
> Sorry to select you out @eileencodes but I have seen you being active in the PRs so I thought maybe you could take a look at this and point out any things we should change to get this merged.

Please don't do this. I get hundreds of emails a day and I can only look at so many PRs. Sometimes PRs take time to get merged or get lost because the core team is small for the number of changes that come to Rails.

erichmachado commented:
Rebased, fixed the conflicts, refactored the code, changed the default setting for new/upgraded apps and updated the documentation accordingly. Hopefully this is good to be merged now @eugeneius. 😅 🤞

james-caresnap commented:
@erichmachado thanks for keeping this PR alive. It's unfortunate that it has not been merged even though it's been open for more than a year and a half. I'd like to express my support for getting this bug fixed.

erichmachado commented:
> @erichmachado thanks for keeping this PR alive. It's unfortunate that it has not been merged even though it's been open for more than a year and a half. I'd like to express my support for getting this bug fixed.

Thank you for your support @james-caresnap. I'll find some time later this week to update this and get it ready to merge again.

We'll probably have to post this onto the [official mailing list](https://discuss.rubyonrails.org/c/rubyonrails-core/5) to get some traction. Would you like to help me with that?

james-caresnap commented:
Happy to help.

On Wed, Jul 28, 2021 at 9:57 AM Erich Soares Machado <
***@***.***> wrote:

> @erichmachado <https://github.com/erichmachado> thanks for keeping this
> PR alive. It's unfortunate that it has not been merged even though it's
> been open for more than a year and a half. I'd like to express my support
> for getting this bug fixed.
>
> Thank you for your support @james-caresnap
> <https://github.com/james-caresnap>. I'll find some time later this week
> to update this and get it ready to merge again.
>
> We'll probably have to post this onto the official mailing list
> <https://discuss.rubyonrails.org/c/rubyonrails-core/5> to get some
> traction. Would you like to help me with that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/37682#issuecomment-888380078>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APPKXZ7K6QYGE5MFUSYTHRDT2ALG7ANCNFSM4JLP7VJQ>
> .
>

erichmachado commented:
> Happy to help.
> […](#)

Created [a topic](https://discuss.rubyonrails.org/t/namespaced-uuid-generation-bug-fix/78569) on the official mailing list @james-caresnap.

eileencodes reviewed (changes requested):
Sorry for the wait, @erichmachado. A couple things need to be fixed and then I can merge this.

zzak commented on railties/test/application/configuration_test.rb:
```diff
@@ -2422,6 +2422,20 @@ def index
+    end
+
+    test "ActiveSupport.use_rfc4122_namespaced_uuids is disabled by default for upgraded apps" do
+      remove_from_config '.*config\.load_defaults.*\n'
```
It looks like this kind of test is used earlier in this file, but I find it weird that this test removes the entire `config.load_defaults` as that is not particular to the feature we are testing here.

Would it make more sense explicitly remove (or change) just this setting in a generated app?
